[GeoMechanicsApplication] Extract a static utility function for the calculation of the Permeability Matrix (H)

markelov208 opened the pull request:
**📝 Description**
A brief description of the PR.

- Extraction of a calculation of the permeability matrix as an utility function.

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- created GeoTransportEquationUtilities class with static functions
- created unit tests for the functions
- replaced calculations of the permeability matrix in elements with the utility function calls
- created README.md file with the permeability matrix definition

**Questionable things**
- name GeoTransportEquationUtilities
- name PreparePermeabilityMatrixHForIntegration, it seems it should be used for other matrices.
- bounded_matrix vs matrix

Commit:
extracted utility function for permeability matrix H and tests

Commit:
added utility function for permeability matrix H

Commit:
removed commented lines

Commit:
removed PDimMatrix from input parameters

Commit:
used utility function for permeability matrix in CalculatePermeabilityFlow

Commit:
removed rPDimMatrix from input parameters of CalculatePermeabilityFlow

Commit:
added README.md with the permeability matrix's description

Commit:
const ref on Variables to remove a code smell

Commit:
corrected text in README.md

Commit:
removed unused include statements

Commit:
cleaned test

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,68 @@
+    using VectorType = Vector;
+    using MatrixType = Matrix;
+
+    //----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
Could you remove this `//-------` line? In my opinion it doesn't really add too much clarity.

markelov208 replied:
Removed. The line means nothing. ;(

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/README.md:
```diff
@@ -0,0 +1,27 @@
+File transport_equation_utilities.hpp includes 
+
+-  CalculatePermeabilityMatrixH functions with input parameters defined with bounded matrices or matrices
+-  PreparePermeabilityMatrixHForIntegration function that multiplies the permeabiity matrix for future integration
```
Minor typo
```suggestion
-  PreparePermeabilityMatrixHForIntegration function that multiplies the permeability matrix for future integration
```

markelov208 replied:
Thank you for the catch.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1321,7 +1321,12 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateAndAddPermeabilityMatrix(M
 {
     KRATOS_TRY
 
-    this->CalculatePermeabilityMatrix(rVariables.PDimMatrix, rVariables.PPMatrix, rVariables);
```
If I'm correct, the original `CalculatePermeabilityMatrix` of the elements should not be used anymore right? So we can probably remove that one!

markelov208 replied:
Oh, I removed the call but the original body was left. Many thanks.

avdg81 replied:
Nice cleanup :-)

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1470,20 +1475,20 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateAndAddCompressibilityFlow(
 
-    noalias(rPMatrix) = -PORE_PRESSURE_SIGN_FACTOR * rVariables.DynamicViscosityInverse *
-                        rVariables.RelativePermeability * rVariables.PermeabilityUpdateFactor *
-                        prod(rPDimMatrix, trans(rVariables.GradNpT)) * rVariables.IntegrationCoefficient;
```
Nice catch that this was again the same code

markelov208 replied:
There are many surprises. ;)

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,68 @@
+                            prod(Matrix(prod(GradNpT, PermeabilityMatrix)), trans(GradNpT)) * IntegrationCoefficient;
+    }
+
+    static inline Matrix CalculatePermeabilityMatrixH(const Matrix& GradNpT,
```
Right now, this function calculates the contribution to the permeability matrix of one integration point. If I understand the issue correctly (see e.g. the line in the issue description 'which takes the gradient of the pressure shape function for all integration points, etc'), it should take a list of inputs for all integration points (so probably something like a std::vector<Matrix> GradNpTs, std::vector<double> DynamicViscosities, etc and add all contributions to the total permeability matrix.

It might be a bit more complex to integrate this into the element, because the loop over integration points is higher up in the `CalculateAll` function, but I think it would make it more clear in the end (and as far as I understood, that was also how we discussed in in our coming up with the design).

Please correct me if I'm wrong (also a question to @WPK4FEM and @avdg81)

The challenge with implementing this, as I see it, is that the `CalculateAndAddLHS` and `CalculateAndAddRHS` functions are also defined per integration point instead of for the entire element at once. We need to change that behavior as well in this case, so maybe we can have a discussion about this (since we could also make a small issue that refactors the LHS/RHS functions such that they work on the entire element instead of doing it per integration point)?

markelov208 replied:
As far as understand only Integration coefficient blocks it because it is calculated for each point. Of course, it can be moved  separately and kept as a matrix. Looking forward to @WPK4FEM and @avdg81 suggestion.

avdg81 replied:
Independently of @rfaasse I had the same thoughts about letting this function return the matrix for an element, rather than an integration point contribution. I would suggest to make that change in this PR too. Let's sit together and explore how that can be done.

markelov208 replied:
Yesterday Anne and me had a conversation on this. The decision is to keep the current approach and to make changes later.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    template <unsigned int TNumNodes>
+    static inline void PreparePermeabilityMatrixHForIntegration(BoundedMatrix<double, TNumNodes, TNumNodes>& rPMatrix,
```
If I remember correctly, we discussed that the relative permeability and permeability update factor could be incorporated already in the `const Matrix& PermeabilityMatrix` of the previous function, such that the utility doesn't need to know about it. This is I think what the addition '(which implicitly includes relative permeability and permeability update factor)' in the original issue meant (also here, please correct me if I'm wrong @WPK4FEM and @avdg81)

markelov208 replied:
These variables can be moved back. As far as I remember than it will be not really the permeability matrix. As well it seems this function should have another name because it could be used for another matrices. Looking forward to comments from Wijtze-Pieter and Anne.

avdg81 replied:
The only thing this function and its overload do is to multiply a matrix by a scalar (which is the product of three other scalars). Therefore, extracting such a function only makes sense if the name of that function clarifies the meaning of the resulting matrix-scalar product. I'm not sure whether that applies here.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,68 @@
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Ignasi de Pouplana,
+//                   Vahid Galavi
```
For these new classes, in my opinion we don't need to add these names, so you can just keep only your own there.

markelov208 replied:
OK

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+    // clang-format on
+
+    BoundedMatrix<double, 3, 3> PermeabilityMatrix      = ZeroMatrix(3, 3);
+    double                      DynamicViscosityInverse = 1;
```
```suggestion
    const double                DynamicViscosityInverse = 1;
```

markelov208 replied:
Thank you for pointing. It is done for all double variables below.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+    // clang-format off
+    MaterialPermeabilityMatrix <<= 1.0, 2.0, 3.0, 4.0;
+    // clang-format on
+    double IntegrationCoefficient = 1.0;
```
Can be const 
```suggestion
    const double IntegrationCoefficient = 1.0;
```

markelov208 replied:
done

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+    double                      DynamicViscosityInverse = 1;
+    BoundedMatrix<double, 2, 2> MaterialPermeabilityMatrix;
+    // clang-format off
+    MaterialPermeabilityMatrix <<= 1.0, 2.0, 3.0, 4.0;
```
This could have similar (2x2) formatting as the GradNpT (line 28)

markelov208 replied:
Indeed, changed.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+    noalias(PDimMatrix) = -PORE_PRESSURE_SIGN_FACTOR * prod(GradNpT, MaterialPermeabilityMatrix);
+
+    double RelativePermeability     = 0.02;
+    double PermeabilityUpdateFactor = 1.5;
```
Can be const
```suggestion
    const double RelativePermeability     = 0.02;
    const double PermeabilityUpdateFactor = 1.5;
```

markelov208 replied:
done

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+
+    double RelativePermeability     = 0.02;
+    double PermeabilityUpdateFactor = 1.5;
+    noalias(PMatrix) = DynamicViscosityInverse * RelativePermeability * PermeabilityUpdateFactor *
```
I'm not sure if we need to repeat the same calculation in the UT, we could just check against the numbers we know should come out of the calculation (e.g. by checking by hand once)

markelov208 replied:
The calculation is replaced with numbers.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+    // clang-format on
+
+    BoundedMatrix<double, 4, 4> PermeabilityMatrix      = ZeroMatrix(4, 4);
+    double                      DynamicViscosityInverse = 1;
```
Can be const 
```suggestion
    const double                DynamicViscosityInverse = 1;
```

markelov208 replied:
done

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+                           4.0, 5.0, 6.0,
+                           7.0, 8.0, 9.0;
+    // clang-format on
+    double IntegrationCoefficient = 1.0;
```
```suggestion
    const double IntegrationCoefficient = 1.0;
```

markelov208 replied:
done

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_permeability_matrix.cpp:
```diff
@@ -0,0 +1,109 @@
+    noalias(PDimMatrix) = -PORE_PRESSURE_SIGN_FACTOR * prod(GradNpT, MaterialPermeabilityMatrix);
+
+    double RelativePermeability     = 0.1;
+    double PermeabilityUpdateFactor = 2.0;
```
```suggestion
    const double RelativePermeability     = 0.1;
    const double PermeabilityUpdateFactor = 2.0;
```

markelov208 replied:
done

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,68 @@
+
+    //----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+    template <unsigned int TDim, unsigned int TNumNodes>
+    static inline void CalculatePermeabilityMatrixH(BoundedMatrix<double, TNumNodes, TNumNodes>& rPMatrix,
```
I think it might be possible to make the MatrixType the template argument here, such that both BoundedMatrix and Matrix can be handled in a single function. @avdg81 did something similar for the element utilities (see https://github.com/KratosMultiphysics/Kratos/blob/f8087c739d57934b07a1c7114054fba58909e1fe/applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp#L600-L611

Not sure, but we might be able to have a similar approach here.

avdg81 replied:
In addition to the above suggestion, I'd like to return the function's result using the return type (like you did for the overload) rather than through an output parameter.

Furthermore, I would leave out `H` of the function name (i.e. change it to `CalculatePermeabilityMatrix`), since in other places we also don't include the symbols of the matrices (e.g. `CalculateMassMatrix` rather than `CalculateMassMatrixM`).

markelov208 replied:
There are still two functions; however, only one body. There is a question on functionality. Anne suggested to test later efficiency of BoundedMatrix vs Matrix. I have found a link which discussed this topic but the info can be very outdated. In a few words, there was no difference between BoundedMatrix and Matrix for a large matrices. 

The function name has lost H and it returns a matrix now.

rfaasse reviewed (commented):
Thank you for implementing, unit testing and documenting these new utility functions. It removes a lot of duplication and cleans up the code!

I have some more fundamental questions about the signature of the function, which now calculates contributions per integration point, while in the meeting, I think we discussed the matrix would be calculated for all integration points at once (i.e. the summation/integration is done within the utility function). Looking at the way the RHS/LHS calculations are set-up now in the element itself, this will need extra steps, so I'd propose we have a discussion about this with @avdg81 and @WPK4FEM.

Commit:
fixed some of the review comments

markelov208 reviewed (commented):
Richard, thank you very much for your thorough review. I have fixed small things. The main question as you pointed is at which level to put the calculation. Currently, everything is done for each point. 
Thank you for the link on Anne's modifications. I will take a look. I have kept different inputs because my current understanding is that BoundedMatrix is more efficient than Matrix. I have found an old link when a guy showed a speed up of a factor of 3. Moreover, Matrix is used only in SmallStrainUPwDiffOrder element. Is one of our goals to convert  Matrices into BoundedMatrix?
Looking forward to your, Anne and Wijtze-Pieter comments. 
Many thanks.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,65 @@
+{
+public:
+    using VectorType = Vector;
+    using MatrixType = Matrix;
```
I haven't found any usages of these aliases, so I would suggest to remove them.

markelov208 replied:
Thank you. Removed.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,65 @@
+    static inline void CalculatePermeabilityMatrixH(BoundedMatrix<double, TNumNodes, TNumNodes>& rPMatrix,
+                                                    const Matrix& GradNpT,
+                                                    double        DynamicViscosityInverse,
+                                                    const BoundedMatrix<double, TDim, TDim>& PermeabilityMatrix,
```
Perhaps we should discuss with @WPK4FEM about a better name for this function parameter. The function is named `CalculatePermeabilityMatrix`, but one of its parameters (not being the output parameter) is called `PermeabilityMatrix`. This is rather confusing to me.

markelov208 replied:
Wijtze-Pieter agreed to call it as MaterialPermeabilityMatrix.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/README.md:
```diff
@@ -0,0 +1,27 @@
+
+The mathematical definition of the permeability matrix is:
+$$H = \int_\Omega (\nabla N_p)^T \frac{1}{\mu} k \nabla N_p d\Omega$$
+where $\nabla N_p$ is the gradient of the pressure shape function, $\mu$ is the dynamic viscosity (material parameter) and $k$ is permeability (material parameter)
```
Perhaps we should elaborate a bit about the meaning and appearance of $k$. Initially, I thought that "the" permeability was a scalar, but closer inspection and after some discussion I realized that it is represented by a matrix (to account of orthotropic or anisotropic permeability?).

markelov208 replied:
The description is updated.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/README.md:
```diff
@@ -0,0 +1,27 @@
+
+File transport_equation_utilities.hpp includes 
+
+-  CalculatePermeabilityMatrixH functions with input parameters defined with bounded matrices or matrices
```
When we are able to merge these two implementations, this remark is no longer needed.

markelov208 replied:
I have removed the remark to avoid changes in future. ;)

avdg81 reviewed (commented):
This is a very nice PR where the repetition of how the permeability is calculated is eliminated. Well done. I would suggest to take this one step further: let's move the summation of the integration point contributions to inside your extracted function.

Commit:
changes to answer the review comments

- merging PreparePermeabilityMatrixHForIntegration into CalculatePermeabilityMatrix
- updating tests
- updating README file
- etc.

markelov208 reviewed (commented):
Richard and Anne, many thanks for the comments and discussion. Hope we can move further to other issues. ;)

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -0,0 +1,53 @@
+                                           PermeabilityUpdateFactor, IntegrationCoefficient);
+    }
+
+    static inline Matrix CalculatePermeabilityMatrix(const Matrix& rGradNpT,
+                                                     double        DynamicViscosityInverse,
+                                                     const Matrix& rMaterialPermeabilityMatrix,
+                                                     double        RelativePermeability,
+                                                     double        PermeabilityUpdateFactor,
+                                                     double        IntegrationCoefficient)
+    {
+        return -PORE_PRESSURE_SIGN_FACTOR * DynamicViscosityInverse *
+               prod(rGradNpT, Matrix(prod(rMaterialPermeabilityMatrix, trans(rGradNpT)))) *
+               RelativePermeability * PermeabilityUpdateFactor * IntegrationCoefficient;
+    }
```
Since this is not a template function, I think we can create a definition and implementation (and move the implementation to a .cpp file).

rfaasse reviewed (dismissed):
Thanks for incorporating the comments, this cleaned up quite a bit of duplication and is nicely unit tested. I have one minor comment to move an implementation to a .cpp file, but in my opinion it's not blocking (could also be done in the PR of one of the next matrix calculations), so approving and leaving it up to you!

Commit:
Merge branch 'master' into geo/12211-extract-utility-function-H

Commit:
corrected a small value: 1.1e-16 instead of 1.66e-16

markelov208 reviewed (commented):
Richard and Anne, finally all pipelines are green.

rfaasse reviewed (approved):
Nice, looks good to go for me!
